ocaml-lsp-server/src/semantic_highlighting.ml

```diff
@@ -10,142 +10,94 @@ open Fiber.O
-
-  val variable : t
-
-  val property : t
-
-  val enum_member : t
-
-  val event : t
-
-  val function_ : t
-
-  val method_ : t
-
-  val macro : t
-
-  val keyword : t
-
-  val modifier : t
-
-  val comment : t
-
-  val string : t
-
-  val number : t
-
-  val regexp : t
-
-  val operator : t
-
-  val decorator : t
-
-  val list : string list
+  val to_int : t -> int
 
   val to_legend : t -> string
-end [@warning "-32"] = struct
-  type t = int
-
-  let module_ = 0
-  (* the number is determined by the index in [list] defined below; e.g., we use
-     [0] here because we'd like to use "namespace" token type for module
-     symbols *)
-
-  let type_ = 1
-
-  let class_ = 2
-
-  let enum = 3
-
-  let module_type = 4
-
-  let record = 5
-
-  let type_parameter = 6
-
-  let parameter = 7
-
-  let variable = 8
-
-  let property = 9
 
-  let enum_member = 10
-
-  let event = 11
-
-  let function_ = 12
-
-  let method_ = 13
-
-  let macro = 14
-
-  let keyword = 15
-
-  let modifier = 16
-
-  let comment = 17
-
-  let string = 18
-
-  let number = 19
-
-  let regexp = 20
-
-  let operator = 21
-
-  let decorator = 22
-
-  let list =
-    [ "namespace"
-    ; "type"
-    ; "class"
-    ; "enum"
-    ; "interface"
-    ; "struct"
-    ; "typeParameter"
-    ; "parameter"
-    ; "variable"
-    ; "property"
-    ; "enumMember"
-    ; "event"
-    ; "function"
-    ; "method"
-    ; "macro"
-    ; "keyword"
-    ; "modifier"
-    ; "comment"
-    ; "string"
-    ; "number"
-    ; "regexp"
-    ; "operator"
-    ; "decorator"
+  val tokenTypes : string list
+end = struct
+  type t =
+    | Builtin of SemanticTokenTypes.t
+    | Module
+    | Module_type
+
+  let of_builtin t = Builtin t
+
+  let module_ = Module
+
+  let module_type = Module_type
+
+  let legend : SemanticTokenTypes.t list =
+    [ Namespace
+    ; Type
+    ; Class
+    ; Enum
+    ; Interface
+    ; Struct
+    ; TypeParameter
+    ; Parameter
+    ; Variable
+    ; Property
+    ; EnumMember
+    ; Event
+    ; Function
+    ; Method
+    ; Macro
+    ; Keyword
+    ; Modifier
+    ; Comment
+    ; String
+    ; Number
+    ; Regexp
+    ; Operator
+    ; Decorator
     ]
 
-  let array = lazy (Array.of_list list)
-
-  let to_legend t = (Lazy.force array).(t)
+  let tokenTypes : string list =
+    List.map legend ~f:(fun s ->
+        match SemanticTokenTypes.yojson_of_t s with
+        | `String s -> s
+        | _ -> assert false)
+
+  let simplify : t -> SemanticTokenTypes.t = function
+    | Module -> Namespace
+    | Module_type -> Interface
+    | Builtin s -> s
+
+  let to_int =
+    let module Table = MoreLabels.Hashtbl in
+    let table =
+      lazy
+        (let t = Table.create (List.length legend) in
+         List.iteri legend ~f:(fun data key -> Table.add t ~key ~data);
+         t)
+    in
+    fun t ->
+      Table.find
+        (Lazy.force table)
+        (match t with
+        | Module -> Namespace
+        | Module_type -> Interface
+        | Builtin s -> s)
```

Comment:
This could be `(simplify t)`

Reply:
Done. I simplified the representation a bit more.